### PR TITLE
Add linter for a single element for loop

### LINKED ISSUE #1540: Lint for iterating over a slice with one (or zero) element
Another day a see something like:

```rust
for a in &[b] {
}
```

### PR DESCRIPTION
**Pull request by patrickelectric**
changelog: Fixes #1540, check for vectors that contain a single element in a for loop

### CODE REVIEW

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -1684,6 +1711,27 @@ fn check_for_loop_over_map_kv<'tcx>(
...
+    _: &'tcx Expr<'_>,
+) {
+    if let ExprKind::AddrOf(BorrowKind::Ref, _, ref expr) = arg.kind {
+        if let ExprKind::Array(ref expr_list) = expr.kind {
+            if expr_list.len() == 1 {
+                span_lint(
+                    cx,
+                    SINGLE_ELEMENT_LOOP,
+                    expr.span,
+                    "Do not iterate over a single element array.",
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply clippy_lints/src/loops.rs]: thanks!

**Comment by patrickelectric** [comment]
ping @flip1995

**Comment by patrickelectric** [comment]
ping @flip1995

**Comment by flip1995** [comment]
Sorry for taking so long. I currently have 15 PRs I'm actively reviewing/triaging besides doing other organizational stuff for Clippy. I usually review on weekends, so pinging doesn't really help to speed up my reviews.

File: tests/ui/single_element_loop.rs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Tests from for_loop.rs that don't have suggestions
+
+#[warn(clippy::single_element_loop)]
```
Comment (flip1995) [inline tests/ui/single_element_loop.rs:3]: We should definitely add a suggestion here. Let me know if you need help with that. You can take a look at `span_lint_and_sugg` usages to see how to do this.

File: clippy_lints/src/loops.rs [hunk 3]
```diff
@@ -1684,6 +1711,27 @@ fn check_for_loop_over_map_kv<'tcx>(
     }
 }
 
+fn check_for_single_element_loop<'tcx>(
+    cx: &LateContext<'tcx>,
+    _: &'tcx Pat<'_>,
+    arg: &'tcx Expr<'_>,
+    _: &'tcx Expr<'_>,
+    _: &'tcx Expr<'_>,
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:1719]: No need to add all arguments, if they are unused.

File: clippy_lints/src/loops.rs [hunk 4]
```diff
@@ -453,6 +453,31 @@ declare_clippy_lint! {
...
+    /// }
+    /// ```
+    /// could be written as
+    /// ```rust
+    /// let item1 = 2;
+    /// let item = &item1;
+    /// println!("{}", item);
+    /// ```
+    pub SINGLE_ELEMENT_LOOP,
+    style,
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:477]:
```suggestion
    complexity,
```

**Review by flip1995 — COMMENTED** [review]
Thanks for the contribution and welcome to Clippy!

File: tests/ui/single_element_loop.rs (hunk 2 above)
Comment (patrickelectric) [reply tests/ui/single_element_loop.rs:3]: Thanks for the suggestion, now it's using `span_lint_and_sugg`

File: clippy_lints/src/loops.rs (hunk 3 above)
Comment (patrickelectric) [reply clippy_lints/src/loops.rs]: Oh, I was just following the style of the other functions, removed!

File: clippy_lints/src/loops.rs (hunk 4 above)
Comment (patrickelectric) [reply clippy_lints/src/loops.rs]: Done!

**Comment by patrickelectric** [comment]
Hi @flip1995, sorry for the ping, is just something that I do after a couple of days if a PR does not get a review, but since you mainly review PRs during the weekends I'll stop ping here. Sorry for the bothering! 
I just fixed the requested changes and rebased over the current master branch, let me know if anything else is necessary.
Have a great weekend!

File: clippy_lints/src/loops.rs [hunk 5]
```diff
@@ -1865,6 +1893,29 @@ fn check_for_loop_over_map_kv<'tcx>(
     }
 }
 
+fn check_for_single_element_loop<'tcx>(cx: &LateContext<'tcx>, pat: &'tcx Pat<'_>, arg: &'tcx Expr<'_>) {
+    if_chain! {
+        if let ExprKind::AddrOf(BorrowKind::Ref, _, ref expr) = arg.kind;
+        if let PatKind::Binding(.., target, _) = pat.kind;
+        if let ExprKind::Array(ref expr_list) = expr.kind;
+        if expr_list.len() == 1;
+        if let ExprKind::Path(ref list_item) = expr_list[0].kind;
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:1902]:
```suggestion
        if let [expr] = expr_list; // or a more fitting name than "expr"
        if let ExprKind::Path(ref list_item) = expr.kind;
```

File: clippy_lints/src/loops.rs [hunk 6]
```diff
@@ -1865,6 +1893,29 @@ fn check_for_loop_over_map_kv<'tcx>(
...
+        if let Some(list_tem_name) = single_segment_path(list_item).map(|ps| ps.ident.name);
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:1903]:
```suggestion
        if let Some(list_item_name) = single_segment_path(list_item).map(|ps| ps.ident.name);
```

File: clippy_lints/src/loops.rs [hunk 7]
```diff
@@ -1865,6 +1893,29 @@ fn check_for_loop_over_map_kv<'tcx>(
...
+
+        then {
+            span_lint_and_sugg(
+                cx,
+                SINGLE_ELEMENT_LOOP,
+                expr.span,
+                "do not iterate over a single element array",
+                "try",
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:1911]:
```suggestion
                "`for` loop over a single element",
                "try",
```

File: clippy_lints/src/loops.rs [hunk 8]
```diff
@@ -1865,6 +1893,29 @@ fn check_for_loop_over_map_kv<'tcx>(
...
+        if let ExprKind::Array(ref expr_list) = expr.kind;
+        if expr_list.len() == 1;
+        if let ExprKind::Path(ref list_item) = expr_list[0].kind;
+        if let Some(list_tem_name) = single_segment_path(list_item).map(|ps| ps.ident.name);
+
+        then {
+            span_lint_and_sugg(
+                cx,
+                SINGLE_ELEMENT_LOOP,
+                expr.span,
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:1909]: You should use the span from the start of the for loop to the end of `expr`. You can do this, by also passing the `expr` from the `check_for_loop` function and then use the span `for_expr.span.with_hi(expr.span.hi())`.

File: tests/ui/single_element_loop.rs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+// Tests from for_loop.rs that don't have suggestions
```
Comment (flip1995) [inline tests/ui/single_element_loop.rs:1]:
```suggestion
// run-rustfix
// Tests from for_loop.rs that don't have suggestions
```

**Review by flip1995 — CHANGES_REQUESTED** [review]

**Comment by patrickelectric** [comment]
Needs https://github.com/rust-lang/rust-clippy/pull/6191 to compile.

**Comment by patrickelectric** [comment]
ping @flip1995

**Review by flip1995 — APPROVED** [review]

**Comment by flip1995** [comment]
@bors r+

Thanks!

Sorry for taking so long. I'll start a new job in a new town next month and have little time for reviews currently. I hope it settles down soon and I have more time and energy for Clippy again.

**Comment by patrickelectric** [comment]
Hi @flip1995, thank you for the review, I was just a bit worried about the hacktoberfest schedule.
Congratulations for the new job and I hope all is well in the coming weeks!

**Comment by flip1995** [comment]
IIUC for hacktoberfest only the opening date of the PR is relevant. :wink:

**Pull request merged**